File: C++/binary-search.cpp
#include <vector>


class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        // Time:  O(logn)
        // Space: O(1)
        if (nums.empty()) return -1;
        int start = 0;
        int end = nums.size();
        while (start + 1 < end) {
            int m = start + (end - start) / 2;
            if (nums[m] > target) {
                end = m;
            } else {
                start = m;
            }
        }
        return nums[start] == target? start : -1;
    }
};

File: Python/binary-search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Time:  O(logn)
        # Space: O(1)
        if not nums:
            return -1
        start = 0
        end = len(nums)
        while (start + 1 < end):
            m = start + (end - start) // 2
            if target < nums[m]:
                end = m
            else:
                start = m
        return start if nums[start] == target else -1
